### PR TITLE
feat(registration): create Registration feature project

### PR DESCRIPTION
**Pull request by zbaylin**
This is a WIP so far, but my plan is to eventually have it so that:
- There is a Command Palette entry entitled `Enter License Key`
- There is an icon in the status bar which shows registration status (similar to Sublime Text, although in the status bar rather than the title bar)
- Clicking on the icon or selecting the command palette entry will open the quick menu to enter a license key which will use the persistent store

### CODE REVIEW

**Comment by zbaylin** [comment]
@bryphe I was finally able to finish this up -- sorry for the delay!

On Windows and macOS, there are icons/text in the titlebar that one can click to open the registration panel:
image

Clicking on that icon or using the quick menu brings up this menu:
image

Entering the license key will check with the API endpoint, and will persist using the `Global` persistence store in `Store/Persistence.re`.

Hopefully the code change is good! There were definitely some new things I was working with, so let me know if anything is "off"!

File: src/Store/Persistence.re [hunk 1]
```diff
@@ -26,14 +30,20 @@ module Global = {
     lazy(
       {
         instantiate("global", () =>
-          Schema.[entry(version), entry(workspace), entry(extensionValues)]
+          Schema.[
+            entry(version),
+            entry(workspace),
+            entry(extensionValues),
+            entry(licenseKey),
```
Comment (bryphe) [inline src/Store/Persistence.re:37]: Thanks for moving this to persistence! :+1:

File: src/Model/FocusManager.re [hunk 2]
```diff
@@ -27,6 +27,8 @@ let current = (state: State.t) =>
     Focus.LanguageSupport;
   } else if (Feature_Registers.isActive(state.registers)) {
     Focus.InsertRegister;
+  } else if (Feature_Registration.isActive(state.registration)) {
+    Focus.LicenseKey;
```
Comment (bryphe) [inline src/Model/FocusManager.re:31]: Cool, thanks for adding the new focus state here - looks good!

File: src/Feature/Registration/Feature_Registration.re [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+          Service_Registration.Effect.checkLicenseKeyValidity(licenseKey)
+          |> Isolinear.Effect.map(wrapEvent),
+        );
+      | _ => Nothing
+      };
+
+    (model', outmsg);
+  | Pasted(text) =>
+    let inputModel = Component_InputText.paste(~text, model.inputModel);
+    ({...model, inputModel}, Nothing);
```
Comment (bryphe) [inline src/Feature/Registration/Feature_Registration.re:122]: I tested on Linux - and for some reason the `Paste` event wasn't working. I think some of the context keys might not be triggered - https://github.com/onivim/oni2/blob/92971459a0529f398f6f6cea1859511adacb10e8/src/Store/KeyBindingsStoreConnector.re#L71

I'd expect for this we'd have the `textInputFocus` context key active.

File: src/Feature/Registration/Feature_Registration.re [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+                icon=FontAwesome.checkCircle
+                fontSize=14.
+                color={Colors.foreground.from(theme)}
+              />
+            </View>
+            <Text
+              fontFamily={font.family}
+              fontSize=14.
+              style=Styles.messageText
+              text="Thank you for purchasing Onivim!"
```
Comment (bryphe) [inline src/Feature/Registration/Feature_Registration.re:287]: I like the flow you put together for this! Thanks for thinking through the UX here

File: src/Feature/Registration/Feature_Registration.re (hunk 3 above)
Comment (bryphe) [reply src/Feature/Registration/Feature_Registration.re:122]: Just pushed up a fix for this, to expose the context keys

**Review by bryphe — APPROVED** [review]
Looks great, @zbaylin ! Thanks for your work on this - with this, it seems like we're in good shape for auto-updates! :100:

**Comment by zbaylin** [comment]
Awesome @bryphe! Thanks for that context key fix! Can't wait to see this interact with #2430 and #2497!

**Comment by zbaylin** [comment]
I think this PR is haunted with transient CI failures 😄

**Comment by bryphe** [comment]
> I think this PR is haunted with transient CI failures smile

Sorry about that, not really your PR - all of them have been! I'm working through them here:
- https://github.com/onivim/oni2/pull/2678
- https://github.com/onivim/oni2/pull/2681
- https://github.com/onivim/oni2/pull/2682

...hopefully in better shape now. To make matters worse - because of those combined intermittences, we no longer have cached builds (they expired) - so CI has been very slow because it's always rebuilding the world! :crossed_fingers: it'll be better shortly once those PRs percolate into master.

**Pull request merged**